app/src/main/java/free/rm/skytube/gui/businessobjects/CleanerDialog.java

```diff
@@ -0,0 +1,131 @@
+    private Single<Pair<Integer,Long>> cleanWatchedDownloads(@NonNull Context context) {
+        return Single.fromCallable(() -> {
+                    // Stats
+                    AtomicInteger removedVideosCnt = new AtomicInteger();
+                    AtomicLong removedB = new AtomicLong();
+                    // Maybe rewrite this with reactivex, so that it's parallel
+                    DownloadedVideosDb.getVideoDownloadsDb().getDownloadedVideos().stream()
+                            .filter((video) -> DownloadedVideosDb.getVideoDownloadsDb().isVideoDownloaded(video.getVideoId()).blockingGet()) // Filter only downloaded videos
+                            .filter((video) -> PlaybackStatusDb.getPlaybackStatusDb().getVideoWatchedStatus(video.getId()).isFullyWatched()) // Filter only watched videos
+                            .forEach((video) -> {
+                                DownloadedVideosDb.Status status = DownloadedVideosDb.getVideoDownloadsDb().getDownloadedFileStatus(context, video.getVideoId()).blockingGet();
+                                // Try to remove download
+                                DownloadedVideosDb.getVideoDownloadsDb().removeDownload(context, video.getVideoId()).subscribe();
+                                // Stats
+                                removedVideosCnt.getAndIncrement();
+                                removedB.addAndGet(status.getLocalVideoFile().length());
+                                removedB.addAndGet(status.getLocalAudioFile() != null ? status.getLocalAudioFile().length() : 0);
+                            });
+                    // Format everything in a nice string to output
+                    return new Pair<>(removedVideosCnt.intValue(), removedB.longValue() / 1000 / 1000 /* in MB */);
+                }).subscribeOn(Schedulers.computation())
+                .observeOn(AndroidSchedulers.mainThread())
+                .doOnError(error -> SkyTubeApp.notifyUserOnError(context, error)
+                ).doOnSuccess(msg -> Logger.i(this, "(Downloads)" + SkyTubeApp.getStr(R.string.cleaner_toast_finish),
+                        msg.first, msg.second));
+    }
+
+    /**
+     * Undo bookmark for watched videos, that are not necessary anymore
+     *
+     * @param context The Context for showing Toasts
+     */
+    private Single<Pair<Integer,Long>> cleanWatchedBookmarks(@NonNull Context context) {
+        return Single.fromCallable(() -> {
+                    // Stats
+                    AtomicInteger removedVideosCnt = new AtomicInteger();
+                    AtomicLong removedB = new AtomicLong();
+                    // Maybe rewrite this with reactivex, so that it's parallel
+                    BookmarksDb.getBookmarksDb().getBookmarkedVideos(BookmarksDb.getBookmarksDb().getTotalBookmarkCount().blockingGet(), null).first.stream()
+                            .filter((video) -> BookmarksDb.getBookmarksDb().isBookmarked(video.getId())) // Filter only bookmarked videos
```

Comment:
Here, similar things happens, getBookmarkedVideos returns a list of bookmarked videos, no need for filtering it again.

Reply:
Jup